external/types/web.d.ts

```diff
@@ -0,0 +1,924 @@
+   * @property {IL10n} l10n - Localization service.
+   */
+
+  interface PDFPageViewOptions {
+    container: HTMLDivElement;
+    eventBus: EventBus;
+    id: PageIndex;
+    scale: number;
+    defaultViewport: PageViewport;
+    renderingQueue: PDFRenderingQueue;
+    textLayerFactory: IPDFTextLayerFactory;
+    annotationLayerFactory: IPDFAnnotationLayerFactory;
+    enhanceTextSelection: boolean;
+    renderInteractiveForms: boolean;
+    renderer: RendererType;
+    l10n: IL10n;
+  }
+
+  class PDFPageView {
+    id: PageIndex;
+    renderingId: string;
+    pdfPage: PDFPageProxy | null;
+    pageLabel: string | null;
+    rotation: number;
+    scale: number;
+    viewport: PageViewport;
+    pdfPageRotate: number;
+    hasRestrictedScaling: boolean;
+    enhanceTextSelection: boolean;
+    renderInteractiveForms: boolean;
+    eventBus: EventBus;
+    renderingQueue: PDFRenderingQueue;
+    textLayerFactory: IPDFTextLayerFactory;
+    annotationLayerFactory: IPDFAnnotationLayerFactory;
+    renderer: RendererType;
+    l10n: IL10n;
+    paintTask: RenderTask | null;
+    paintedViewportMap: WeakMap<HTMLCanvasElement | SVGElement, PageViewport>;
+    renderingState: RenderingStates;
+    resume: () => void | null;
+    error: Error | null;
+    onBeforeDraw: () => void | null;
+    onAfterDraw: () => void | null;
+    annotationLayer: AnnotationLayerBuilder | null;
+    textLayer: TextLayerBuilder | null;
+    zoomLayer: any | null;
+    div: HTMLDivElement;
+
+    constructor(options: PDFPageViewOptions);
+    setPdfPage(pdfPage: PDFPageProxy): void;
+    destroy(): void;
+    reset(keepZoomLayer?: boolean, keepAnnotations?: boolean): void;
+    update(scale?: number, rotation?: number): void;
+    cancelRendering(keepAnnotations?: boolean): void;
+    cssTransform(
+      target: HTMLCanvasElement | SVGElement,
+      redrawAnnotations?: boolean
+    ): void;
+    width: number;
+    height: number;
+    getPagePoint(x: number, y: number): Point;
+    draw(): Promise<void>;
+    paintOnCanvas(canvasWrapper: HTMLDivElement): RenderTask;
+    paintOnSvg(wrapper: HTMLDivElement): RenderTask;
+    /**
+     * @param {string|null} label
+     */
+    setPageLabel(label: string | null): void;
+  }
+
+  //
+  // web/pdf_viewer.js
+  //
+  class PDFViewer extends BaseViewer {
+    update(): void;
+  }
+
+  //
+  // web/pdf_thumbnail_viewer.js
+  //
+
+  /**
+   * @typedef {Object} PDFThumbnailViewerOptions
+   * @property {HTMLDivElement} container - The container for the thumbnail
+   *   elements.
+   * @property {IPDFLinkService} linkService - The navigation/linking service.
+   * @property {PDFRenderingQueue} renderingQueue - The rendering queue object.
+   * @property {IL10n} l10n - Localization service.
+   */
+  interface PDFThumbnailViewerOptions {
+    container: HTMLDivElement;
+    linkService: IPDFLinkService;
+    renderingQueue: PDFRenderingQueue;
+    l10n: IL10n;
+  }
+
+  class PDFThumbnailViewer {
```

Comment:
This isn't exposed, please see https://github.com/mozilla/pdf.js/blob/master/web/pdf_viewer.component.js for a list of the viewer components that are.

Reply:
Right, the only things that are exposed directly are defined in `pdf.d.ts` here: https://github.com/mozilla/pdf.js/pull/9369/files/74331b9822b16f59456667b935b3ea430b85b29c#diff-1fc4703ffd3b1030ba834559462f6879R80

A client can't reference `PDFThumbnailViewer` directly; they will have to get one through an access path. (Well, there is one exception which is through the [`types` namespace](https://github.com/mozilla/pdf.js/pull/9369/files/74331b9822b16f59456667b935b3ea430b85b29c#diff-1fc4703ffd3b1030ba834559462f6879R151), which is just an alias for the internal `_pdfjs` namespace. I added `types` because it can be convenient to declare variables, arguments, and return values using types that are not necessarily exposed directly. There is warning that this is not intended to access non-types, because they won't work.)

Some of the things that *are* exposed are `PDFViewer`, `PDFSinglePageViewer`, and `PDFPageView`, all of which have an access path that reaches to `PDFThumbnailViewer` and `PDFThumbnailView`. For example:

`PDFViewer` => `PDFRenderingQueue`:
https://github.com/mozilla/pdf.js/pull/9369/files/74331b9822b16f59456667b935b3ea430b85b29c#diff-4e0cc7eb4db8e8e1efac0bc43f50e0b5R685
`PDFRenderingQueue` => `PDFThumbnailViewer`:
https://github.com/mozilla/pdf.js/pull/9369/files/74331b9822b16f59456667b935b3ea430b85b29c#diff-4e0cc7eb4db8e8e1efac0bc43f50e0b5R230

I just confirmed dynamically that the field `pdfThumbnailViewer` is accessible (though set to null) in the `simpleviewer` example through the access path: `pdfViewer.renderingQueue.pdfThumbnailViewer`. It's `null`, so *if* `pdfThumbnailViewer` is ever set, it will be accessible dynamically there.

For `PDFViewer` the path is `PDFViewer` => `BaseViewer` => `PDFViewerOptions` => `PDFRenderingQueue`: https://github.com/mozilla/pdf.js/pull/9369/files/74331b9822b16f59456667b935b3ea430b85b29c#diff-4e0cc7eb4db8e8e1efac0bc43f50e0b5R424

`PDFThumbnailView` is only accessible through `PDFThumbnailViewer`, which is only accessible through `PDFRenderingQueue`, so perhaps one of the following is the case:
- `PDFRenderingQueue` is not supposed to be an accessible API?
- `PDFRenderingQueue` is supposed to be opaque, with no accessible internals?
- `PDFThumbnailViewer` is not supposed to be an accessible API, specifically from `PDFRenderingQueue`?
- `PDFThumbnailViewer` is accessible but opaque?
- `PDFThumbnailViewer` actually *is* supposed to be accessible, but only indirectly through access paths that go through `PDFRenderingQueue`?